core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1577,4 +1583,14 @@ protected final void onChannelClosed(Channel channel) {
             }
         }
     }
+
+    /**
+     * Ensures this transport is still started / open
+     * @throws IllegalStateException if the transport is not started / open
+     */
+    protected final void ensureOpen() {
```

Comment:
again, the ensureOpen is only a check to fail if we are closed. Yet, if you run it under the lock you make sure that nobody else is modifying a structure protected by this lock. The state of being open / closed is not protected by this lock. It's a volatile read that's it. I don't understand what your issue is here to be honest. The lock has nothing todo with the ensureOpen call. We do this in many other places and almost never protect that once the check has passed a close can't be concurrently happen. For the check itself it doesn't matter,  it might matter for the caller but the callers context can have this extra protection. In this case here all places where it's used are accidentally protected by a lock but that is not required for this check.